### PR TITLE
Added validations to configure dialog

### PR DESCRIPTION
**Pull request by smartura**
This pull request makes changes to the TDE migration dialog of the Azure SQL Migration extension.
A user is now required to pass validation checks prior to completing the configuration dialog.

Failed validation:
image

Successful validation:
image

### CODE REVIEW

**Review by Ramudaykumar — COMMENTED** [review]
Please add required comments, it will help others understand code.

File: extensions/sql-migration/src/dialog/tdeConfiguration/tdeConfigurationDialog.ts [hunk 1]
```diff
@@ -130,6 +136,24 @@ export class TdeConfigurationDialog {
 			adsMethodButton.onDidChangeCheckedState(async checked => {
 				if (checked) {
 					this.migrationStateModel.tdeMigrationConfig.setPendingTdeMigrationMethod(ConfigDialogSetting.ExportCertificates);
+					let validationTitleData = await this.migrationStateModel.getTdeValidationTitles();
+
+					let networkPathValidated =
+						(this.migrationStateModel.tdeMigrationConfig.getPendingNetworkPath() !== '') &&
```
Comment (Ramudaykumar) [inline extensions/sql-migration/src/dialog/tdeConfiguration/tdeConfigurationDialog.ts:142]: Question: don’t we have any thing like string.empty or whitespace to check in typescript?

File: extensions/sql-migration/src/constants/strings.ts [hunk 2]
```diff
@@ -1515,6 +1515,13 @@ export const TDE_WIZARD_MSG_MANUAL = localize('sql.migration.tde.msg.manual', "Y
 export const TDE_WIZARD_MSG_TDE = localize('sql.migration.tde.msg.tde', "You have given Azure Data Studio access to migrate the encryption certificates and database.");
 export const TDE_WIZARD_MSG_EMPTY = localize('sql.migration.tde.msg.empty', "No encrypted database selected.");
 
+export const TDE_VALIDATION_TITLE = localize('sql.migration.tde.validation.title', "Validation");
+export const TDE_VALIDATION_STATUS_NOT_RAN = localize('sql.migration.tde.validation.status.not.ran', "Not Ran");
```
[the remark this answers is not in the document]
Comment (brian-harris) [reply extensions/sql-migration/src/constants/strings.ts:1519]: +1 I suggest following sentence casing and text change to "Not run" as mentioned above as well.

File: extensions/sql-migration/src/constants/strings.ts [hunk 3]
```diff
@@ -1515,6 +1515,13 @@ export const TDE_WIZARD_MSG_MANUAL = localize('sql.migration.tde.msg.manual', "Y
...
+export const TDE_VALIDATION_STATUS_RUNNING = localize('sql.migration.tde.validation.running', "Running");
+export const TDE_VALIDATION_STATUS_SUCCEEDED = localize('sql.migration.tde.validation.status.succeeded', "Succeeded");
+export const TDE_VALIDATION_REQUIREMENTS_MESSAGE = localize('sql.migration.tde.validation.requirements.message', "In order for certificate migration to succeed, you must meet all of the requirements listed below.\n\nClick \"Run Validation\" to check requirements are met.");
+export const TDE_VALIDATION_STATUS_RUN_VALIDATION = localize('sql.migration.tde.validation.run.validation', "Run Validation");
```
Comment (brian-harris) [inline extensions/sql-migration/src/constants/strings.ts:1523]: same casing change here: "Run validation"

File: extensions/sql-migration/src/constants/strings.ts [hunk 4]
```diff
@@ -1515,6 +1515,13 @@ export const TDE_WIZARD_MSG_MANUAL = localize('sql.migration.tde.msg.manual', "Y
...
+
 export function TDE_MIGRATION_ERROR(message: string): string {
 	return localize('sql.migration.starting.migration.error', "An error occurred while starting the certificate migration: '{0}'", message);
```
Comment (brian-harris) [inline extensions/sql-migration/src/constants/strings.ts:1526]: nit:  "The following error has occurred while starting the certificate migration..."

File: extensions/sql-migration/src/dialog/tdeConfiguration/tdeConfigurationDialog.ts [hunk 5]
```diff
@@ -245,15 +270,175 @@ export class TdeConfigurationDialog {
...
+				inputType: 'text',
+				height: 142,
+				multiline: true,
+				CSSStyles: { 'overflow': 'none auto' }
+			})
+			.component();
+
+		const postValidationSeparator = _view.modelBuilder.separator()
+			.withProps(
+				{
```
Comment (brian-harris) [inline extensions/sql-migration/src/dialog/tdeConfiguration/tdeConfigurationDialog.ts:383]: can this emtpy .withProps call be removed.

File: extensions/sql-migration/src/dialog/tdeConfiguration/tdeConfigurationDialog.ts [hunk 6]
```diff
@@ -245,15 +270,175 @@ export class TdeConfigurationDialog {
...
+						let tips = this._validationSuccessDescriptionErrorAndTips[rowIndex][3];
+
+						message = `Description:${EOL}${description}`;
+						if (!successful) {
+							message += `${EOL}${EOL}`;
+							if (errorMessage?.length > 0) {
+								message += `Error:${EOL}${errorMessage}${EOL}${EOL}`;
+							}
+
+							message += `Troubleshooting Tips:${EOL}${tips}`;
```
Comment (brian-harris) [inline extensions/sql-migration/src/dialog/tdeConfiguration/tdeConfigurationDialog.ts:365]: are the strings "Description:", "Error:", "Troubleshooting Tips:" all user strings?  if so they would need to be added as resource strings for localization

File: extensions/sql-migration/src/constants/strings.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (smartura) [reply extensions/sql-migration/src/constants/strings.ts:1519]: I can lowercase, that's fine.  But shouldn't I keep it in the past tense "Not ran"... as in it was never ran?

File: extensions/sql-migration/src/dialog/tdeConfiguration/tdeConfigurationDialog.ts (hunk 6 above)
Comment (smartura) [reply extensions/sql-migration/src/dialog/tdeConfiguration/tdeConfigurationDialog.ts:365]: Good call, I forgot about these.

File: extensions/sql-migration/src/dialog/tdeConfiguration/tdeConfigurationDialog.ts (hunk 1 above)
Comment (smartura) [reply extensions/sql-migration/src/dialog/tdeConfiguration/tdeConfigurationDialog.ts:142]: I found an empty string constant in utils.ts.  I will use that instead.

File: extensions/sql-migration/src/constants/strings.ts [hunk 7]
```diff
@@ -1515,8 +1515,18 @@ export const TDE_WIZARD_MSG_MANUAL = localize('sql.migration.tde.msg.manual', "Y
...
 
+export const TDE_VALIDATION_TITLE = localize('sql.migration.tde.validation.title', "Validation");
+export const TDE_VALIDATION_REQUIREMENTS_MESSAGE = localize('sql.migration.tde.validation.requirements.message', "In order for certificate migration to succeed, you must meet all of the requirements listed below.\n\nClick \"Run validation\" to check that requirements are met.");
+export const TDE_VALIDATION_STATUS_NOT_RAN = localize('sql.migration.tde.validation.status.not.ran', "Not ran");
+export const TDE_VALIDATION_STATUS_RUNNING = localize('sql.migration.tde.validation.running', "Running");
+export const TDE_VALIDATION_STATUS_SUCCEEDED = localize('sql.migration.tde.validation.status.succeeded', "Succeeded");
+export const TDE_VALIDATION_STATUS_RUN_VALIDATION = localize('sql.migration.tde.validation.run.validation', "Run validation");
+export const TDE_VALIDATION_DESCRIPTION = localize('sql.migration.tde.validation.description', "Description");
+export const TDE_VALIDATION_ERROR = localize('sql.migration.tde.validation.error', "Error");
+export const TDE_VALIDATION_TROUBLESHOOTING_TIPS = localize('sql.migration.tde.validation.troubleshooting.tips', "Troubleshooting Tips");
```
Comment (Ramudaykumar) [inline extensions/sql-migration/src/constants/strings.ts:1526]: Please add same casing here as well.

File: extensions/sql-migration/src/constants/strings.ts [hunk 8]
```diff
@@ -1515,8 +1515,18 @@ export const TDE_WIZARD_MSG_MANUAL = localize('sql.migration.tde.msg.manual', "Y
 export const TDE_WIZARD_MSG_TDE = localize('sql.migration.tde.msg.tde', "You have given Azure Data Studio access to migrate the encryption certificates and database.");
 export const TDE_WIZARD_MSG_EMPTY = localize('sql.migration.tde.msg.empty', "No encrypted database selected.");
 
+export const TDE_VALIDATION_TITLE = localize('sql.migration.tde.validation.title', "Validation");
+export const TDE_VALIDATION_REQUIREMENTS_MESSAGE = localize('sql.migration.tde.validation.requirements.message', "In order for certificate migration to succeed, you must meet all of the requirements listed below.\n\nClick \"Run validation\" to check that requirements are met.");
+export const TDE_VALIDATION_STATUS_NOT_RAN = localize('sql.migration.tde.validation.status.not.ran', "Not ran");
+export const TDE_VALIDATION_STATUS_RUNNING = localize('sql.migration.tde.validation.running', "Running");
```
Comment (brian-harris) [inline extensions/sql-migration/src/constants/strings.ts:1521]: nit:  suggest "Not run" or "Pending"

**Review by brian-harris — APPROVED** [review]

File: extensions/sql-migration/src/constants/strings.ts (hunk 8 above)
Comment (smartura) [reply extensions/sql-migration/src/constants/strings.ts:1521]: I like "Pending" a lot better actually.  Let me make the change.

File: extensions/sql-migration/src/dialog/tdeConfiguration/tdeConfigurationDialog.ts (hunk 1 above)
Comment (smartura) [reply extensions/sql-migration/src/dialog/tdeConfiguration/tdeConfigurationDialog.ts]: Actually, I'm not sure who maintains this utils file, I think it is better not to depend on it.  For now I'll keep the '' string comparison.

Comment (smartura) [reply extensions/sql-migration/src/dialog/tdeConfiguration/tdeConfigurationDialog.ts]: Importing that caused extension to fail to run

**Pull request merged**